services/controller/src/plz/controller/instances/aws/ec2_instance.py

```diff
@@ -35,25 +35,33 @@ def __init__(self,
                  images: Images,
                  containers: Containers,
                  volumes: Volumes,
-                 execution_id: str,
+                 container_execution_id: str,
                  data: dict,
                  redis: StrictRedis):
         super().__init__(redis)
         self.client = client
         self.images = images
         self.delegate = DockerInstance(
-            images, containers, volumes, execution_id, redis)
+            images, containers, volumes, container_execution_id, redis)
         self.data = data
 
-    def run(self,
-            command: List[str],
-            snapshot_id: str,
-            parameters: Parameters,
-            input_stream: Optional[io.BytesIO],
-            docker_run_args: Dict[str, str]):
-        self.images.pull(snapshot_id)
-        self.delegate.run(command, snapshot_id, parameters, input_stream,
-                          docker_run_args)
+    def run_if_free(self,
+                    command: List[str],
+                    snapshot_id: str,
+                    parameters: Parameters,
+                    input_stream: Optional[io.BytesIO],
+                    docker_run_args: Dict[str, str],
+                    max_idle_seconds: int = 0) -> bool:
```

Comment:
I think I'd prefer if this was still named `run` and raised a `ResourceConflictException` or something if it's not free. Too easy to forget to check a boolean.